popup_banners: Redesign connection error banner.

sayamsamal opened the pull request:
This PR serves as the base commit for the Popup banners and redesigns the connection error banner.

Fixes: #31282

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
| Before | After |
|--------|--------|
| ![Screenshot 2025-03-05 at 12 31 53 PM](https://github.com/user-attachments/assets/3e507a03-b36e-4165-b2a2-794e1a22a8aa) | ![Screenshot 2025-03-05 at 12 21 28 PM](https://github.com/user-attachments/assets/ffeb5199-a92a-4a5b-aa12-cac2aeddcae7) | 
| ![Screenshot 2025-03-05 at 12 32 11 PM](https://github.com/user-attachments/assets/575280ad-e8f9-41ae-9138-463266b1ca35) | ![Screenshot 2025-03-05 at 12 20 57 PM](https://github.com/user-attachments/assets/708ee2d7-a49c-4575-b398-a9459667ce8a) | 

**Animation and stacking order:**

https://github.com/user-attachments/assets/bcceba43-126c-44b5-8d24-f2bf86dcdaf9


**Responsiveness:**

https://github.com/user-attachments/assets/08dab3e4-dd3b-4876-87e3-15dbb3e2b8cc


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Hm, the behavior when I press "try now" seems different from `main`? In `main`, the banner disappears for a bit, which makes it feel like the app is doing what you asked. When I test on the PR, the banner doesn't go anywhere, which makes it feel like clicking the app didn't do anything.

It might well actually be doing the same thing in the backend, but I think we also need to clear that banner in response to the click.

alya commented:
(discussed in DMs) The next step here is to add a loading indicator on the button when you press it, so that the user can tell that their click did something.

timabbott commented:
I'm a bit surprised that this state line-wraps the "Try now" button, when there's still plenty of space for it in the first line:

![image](https://github.com/user-attachments/assets/121ac21a-eee9-48fe-9bb7-2ed9e3895c53)

If it's complicated to fix, that can be filed as a follow-up issue.

timabbott commented on web/styles/alerts.css:
```diff
@@ -35,22 +35,33 @@
     overscroll-behavior: contain;
+    /* Set top padding to account for the translate-y motion of the
+       animation to prevent the vertical scroll bar from appearing. */
+    padding-top: var(--popup-banner-translate-y-distance);
```
The `alert-box` class is also used in `templates/zerver/portico.html`.

It might be that the right answer is to sever that -- create  `portico/portico_alerts.css` that's a copy of what we have here for the portico pages, so we can freely work on improving the styling for the app without worrying about breaking that.

@amanagr can help with verifying the updated version doesn't break alert banners on billing/login pages.

timabbott replied:
Hmm, actually, I'm not sure that the portico `alert-box` is not dead code. Maybe worth checking that first, and just deleting the element if so.

timabbott replied:
Or maybe it's fine, it looks like it's just used for `blueslip` error reporting in portico? I think that's confirmed by 4d3daa7ee09365f0945fcef9a4347a1cb5e56b53. So we should just make sure we're not breaking it through adding use of a CSS variable here that won't be set.

sayamsamal replied:
Yep, I had the [prerequisite knowledge](https://chat.zulip.org/#narrow/channel/431-redesign-project/topic/broader.20range.20of.20font.20sizes/near/2107271) that the alert-box was being used for blueslip errors in portico, but overlooked the fact that we don't share `app_variables.css` with portico.

sayamsamal replied:
Adding a `common_variables.css` sounds like a good idea, and I can see it being helpful — I remember an instance where the `--base-font-size-px` was needed in portico, I think it was when I was working on `devtools/buttons`. But in this particular scenario, we would actually want to remove the alert-box related code once we migrate all the banners to the new design, so I think we can do with just defining the `popup-banner-translate-y-distance` in `alerts.css` with a comment to remove it once we migrate.

sayamsamal replied:
I tried emulating the alert popup in portico, and they seem to be working well in regards to the animation.

https://github.com/user-attachments/assets/df5d653f-672d-4e55-a819-cde637db6181

sayamsamal replied:
As you can see from the recording, the close button wasn't working on these alerts in portico. So I have added another commit to fix that at the end.

sayamsamal commented:
## Updates
- Added the loading indicator to the "Try now" button to visually indicate the execution of the retry sequence.

https://github.com/user-attachments/assets/2192845c-26d4-41eb-9188-caa57cd0fce4

- Fixed the other feedback related to `popup-banner-translate-y-distance` in `alerts.css`.
- I think it would be better to tackle the responsiveness of these banners in a separate banner after some discussion at CZO.

alya commented:
Not a blocker for merging, but was this supposed to fix the behavior when the banner disappears/reappears? I still see that happening with network fully disabled.

alya commented:
I don't love the button wrapping behavior @timabbott noted above (https://github.com/zulip/zulip/pull/33823#issuecomment-2701817765).

timabbott commented on web/src/popup_banners.ts:
```diff
@@ -0,0 +1,79 @@
+const CONNECTION_ERROR_POPUP_BANNER: Banner = {
+    intent: "danger",
+    label: $t({
+        defaultMessage: "Unable to connect to Zulip. Retrying soon...",
```
Use the unicode `…`, not three literal `.` characters.

sayamsamal replied:
Fixed, thanks!

sayamsamal commented:
> Not a blocker for merging, but was this supposed to fix the behavior when the banner disappears/reappears? I still see that happening with network fully disabled.

Hm, I checked, and the disappearance of the banner is still indeed present as you noted. It is due to the interplay between the logic for showing/hiding the banner in `web/src/message_fetch.ts` vs `web/src/server_events.js`. Since this not a blocked, I'll try fixing that and put up a separate PR to tackle that bug.

timabbott commented on web/src/popup_banners.ts:
```diff
@@ -24,8 +25,18 @@ export function open_connection_error_popup_banner(opts: {
+            // indicate that the retry sequence was executed again but failed.
+            setTimeout(() => {
+                buttons.hide_button_loading_indicator($retry_connection_button);
+            }, 1000);
```
This is a bit of a quirky way to do things. I don't think it's worse than what we had in `main`, though.

timabbott commented:
Merging this; the above two things are worth continuing to spend attention on over the next few days, but this is definitely better than what we had in `main`.
